modules/apps/data-engine/data-engine-taglib/src/main/java/com/liferay/data/engine/taglib/internal/servlet/taglib/util/DataLayoutTaglibUtil.java

```diff
@@ -258,90 +258,55 @@ public static String resolveModule(String moduleName) {
 	}
 
 	@Activate
-	protected void activate() {
+	protected void activate(BundleContext bundleContext) {
 		_dataLayoutTaglibUtil = this;
-	}
-
-	@Reference(
-		cardinality = ReferenceCardinality.MULTIPLE,
-		policy = ReferencePolicy.DYNAMIC,
-		policyOption = ReferencePolicyOption.GREEDY
-	)
-	protected void addDataDefinitionContentType(
-		DataDefinitionContentType dataDefinitionContentType,
-		Map<String, Object> properties) {
 
-		String contentType = GetterUtil.getString(
-			properties.get("content.type"));
+		_dataLayoutBuilderDefinitionServiceTrackerMap =
+			ServiceTrackerMapFactory.openSingleValueMap(
+				bundleContext, DataLayoutBuilderDefinition.class, null,
+				(serviceReference, emitter) -> {
+					String contentType = GetterUtil.getString(
+						serviceReference.getProperty("content.type"));
 
-		if (Validator.isNull(contentType)) {
-			return;
-		}
-
-		_dataDefinitionContentTypes.put(contentType, dataDefinitionContentType);
-	}
+					if (Validator.isNull(contentType)) {
+						return;
+					}
 
-	@Reference(
-		cardinality = ReferenceCardinality.MULTIPLE,
-		policy = ReferencePolicy.DYNAMIC,
-		policyOption = ReferencePolicyOption.GREEDY
-	)
-	protected void addDataLayoutBuilderDefinition(
-		DataLayoutBuilderDefinition dataLayoutBuilderDefinition,
-		Map<String, Object> properties) {
+					emitter.emit(contentType);
+				});
 
-		String contentType = GetterUtil.getString(
-			properties.get("content.type"));
+		_dataDefinitionContentTypeServiceTrackerMap =
+			ServiceTrackerMapFactory.openSingleValueMap(
+				bundleContext, DataDefinitionContentType.class, null,
+				(serviceReference, emitter) -> {
+					String contentType = GetterUtil.getString(
+						serviceReference.getProperty("content.type"));
 
-		if (Validator.isNull(contentType)) {
-			return;
-		}
+					if (Validator.isNull(contentType)) {
```

Comment:
@JanisZhang Does this logic (if can't get this property, do nothing) match the `openSingleValueMap` method which takes `key` as a parameter?